File: Tests/Tests/STPFileTest.m
//
//  STPFileTest.m
//  Stripe
//
//  Created by Charles Scalesse on 1/8/17.
//  Copyright © 2017 Stripe, Inc. All rights reserved.
//

@import XCTest;



#import "STPTestUtils.h"

@interface STPFile ()

+ (STPFilePurpose)purposeFromString:(NSString *)string;

@end

@interface STPFileTest : XCTestCase

@end

@implementation STPFileTest

#pragma mark - STPFilePurpose Tests

- (void)testPurposeFromString {
    XCTAssertEqual([STPFile purposeFromString:@"dispute_evidence"], STPFilePurposeDisputeEvidence);
    XCTAssertEqual([STPFile purposeFromString:@"DISPUTE_EVIDENCE"], STPFilePurposeDisputeEvidence);

    XCTAssertEqual([STPFile purposeFromString:@"identity_document"], STPFilePurposeIdentityDocument);
    XCTAssertEqual([STPFile purposeFromString:@"IDENTITY_DOCUMENT"], STPFilePurposeIdentityDocument);

    XCTAssertEqual([STPFile purposeFromString:@"unknown"], STPFilePurposeUnknown);
    XCTAssertEqual([STPFile purposeFromString:@"UNKNOWN"], STPFilePurposeUnknown);

    XCTAssertEqual([STPFile purposeFromString:@"garbage"], STPFilePurposeUnknown);
    XCTAssertEqual([STPFile purposeFromString:@"GARBAGE"], STPFilePurposeUnknown);
}

- (void)testStringFromPurpose {
    NSArray<NSNumber *> *values = @[
                                    @(STPFilePurposeDisputeEvidence),
                                    @(STPFilePurposeIdentityDocument),
                                    @(STPFilePurposeUnknown),
                                    ];

    for (NSNumber *purposeNumber in values) {
        STPFilePurpose purpose = (STPFilePurpose)[purposeNumber integerValue];
        NSString *string = [STPFile stringFromPurpose:purpose];

        switch (purpose) {
            case STPFilePurposeDisputeEvidence:
                XCTAssertEqualObjects(string, @"dispute_evidence");
                break;
            case STPFilePurposeIdentityDocument:
                XCTAssertEqualObjects(string, @"identity_document");
                break;
            case STPFilePurposeUnknown:
                XCTAssertNil(string);
                break;
        }
    }
}

#pragma mark - Equality Tests

- (void)testFileEquals {
    STPFile *file1 = [STPFile decodedObjectFromAPIResponse:[STPTestUtils jsonNamed:@"FileUpload"]];
    STPFile *file2 = [STPFile decodedObjectFromAPIResponse:[STPTestUtils jsonNamed:@"FileUpload"]];

    XCTAssertNotEqual(file1, file2);

    XCTAssertEqualObjects(file1, file1);
    XCTAssertEqualObjects(file1, file2);

    XCTAssertEqual(file1.hash, file1.hash);
    XCTAssertEqual(file1.hash, file2.hash);
}

#pragma mark - STPAPIResponseDecodable Tests

- (void)testDecodedObjectFromAPIResponseRequiredFields {
    NSArray<NSString *> *requiredFields = @[
                                            @"id",
                                            @"created",
                                            @"size",
                                            @"purpose",
                                            @"type",
                                            ];

    for (NSString *field in requiredFields) {
        NSMutableDictionary *response = [[STPTestUtils jsonNamed:@"FileUpload"] mutableCopy];
        [response removeObjectForKey:field];

        XCTAssertNil([STPFile decodedObjectFromAPIResponse:response]);
    }

    XCTAssert([STPFile decodedObjectFromAPIResponse:[STPTestUtils jsonNamed:@"FileUpload"]]);
}

- (void)testInitializingFileWithAttributeDictionary {
    NSDictionary *response = [STPTestUtils jsonNamed:@"FileUpload"];
    STPFile *file = [STPFile decodedObjectFromAPIResponse:response];

    XCTAssertEqualObjects(file.fileId, @"file_1AZl0o2eZvKYlo2CoIkwLzfd");
    XCTAssertEqualObjects(file.created, [NSDate dateWithTimeIntervalSince1970:1498674938]);
    XCTAssertEqual(file.purpose, STPFilePurposeDisputeEvidence);
    XCTAssertEqualObjects(file.size, @34478);
    XCTAssertEqualObjects(file.type, @"jpg");

    XCTAssertNotEqual(file.allResponseFields, response);
    XCTAssertEqualObjects(file.allResponseFields, response);
}

@end

File: Stripe/StripeiOSTests/STPFileTest.swift
//  Converted to Swift 5.8.1 by Swiftify v5.8.28463 - https://swiftify.com/
//
//  STPFileTest.m
//  Stripe
//
//  Created by Charles Scalesse on 1/8/17.
//  Copyright © 2017 Stripe, Inc. All rights reserved.
//

@testable import StripePayments
import XCTest

class STPFileTest: XCTestCase {
    // MARK: - STPFilePurpose Tests

    func testPurposeFromString() {
        XCTAssertEqual(STPFile.purpose(from: "dispute_evidence"), .disputeEvidence)
        XCTAssertEqual(STPFile.purpose(from: "DISPUTE_EVIDENCE"), .disputeEvidence)

        XCTAssertEqual(STPFile.purpose(from: "identity_document"), .identityDocument)
        XCTAssertEqual(STPFile.purpose(from: "IDENTITY_DOCUMENT"), .identityDocument)

        XCTAssertEqual(STPFile.purpose(from: "unknown"), .unknown)
        XCTAssertEqual(STPFile.purpose(from: "UNKNOWN"), .unknown)

        XCTAssertEqual(STPFile.purpose(from: "garbage"), .unknown)
        XCTAssertEqual(STPFile.purpose(from: "GARBAGE"), .unknown)
    }

    func testStringFromPurpose() {
        let values: [STPFilePurpose] = [
            .disputeEvidence,
            .identityDocument,
            .unknown,
        ]

        for purpose in values {
            let string = STPFile.string(from: purpose)

            switch purpose {
            case .disputeEvidence:
                XCTAssertEqual(string, "dispute_evidence")
            case .identityDocument:
                XCTAssertEqual(string, "identity_document")
            case .unknown:
                XCTAssertNil(string)
            default:
                break
            }
        }
    }

    // MARK: - Equality Tests

    func testFileEquals() {
        let file1 = STPFile.decodedObject(fromAPIResponse: STPTestUtils.jsonNamed("FileUpload"))
        let file2 = STPFile.decodedObject(fromAPIResponse: STPTestUtils.jsonNamed("FileUpload"))

        XCTAssertEqual(file1, file1)
        XCTAssertEqual(file1, file2)

        XCTAssertEqual(file1?.hash, file1?.hash)
        XCTAssertEqual(file1?.hash, file2?.hash)
    }

    // MARK: - STPAPIResponseDecodable Tests

    func testDecodedObjectFromAPIResponseRequiredFields() {
        let requiredFields = [
            "id",
            "created",
            "size",
            "purpose",
            "type",
        ]

        for field in requiredFields {
            var response = STPTestUtils.jsonNamed("FileUpload")
            response!.removeValue(forKey: field)

            XCTAssertNil(STPFile.decodedObject(fromAPIResponse: response))
        }

        XCTAssertNotNil(STPFile.decodedObject(fromAPIResponse: STPTestUtils.jsonNamed("FileUpload")))
    }

    func testInitializingFileWithAttributeDictionary() {
        let response = STPTestUtils.jsonNamed("FileUpload")!
        let file = STPFile.decodedObject(fromAPIResponse: response)!

        XCTAssertEqual(file.fileId, "file_1AZl0o2eZvKYlo2CoIkwLzfd")
        XCTAssertEqual(file.created, Date(timeIntervalSince1970: 1498674938))
        XCTAssertEqual(file.purpose, .disputeEvidence)
        XCTAssertEqual(file.size, NSNumber(value: 34478))
        XCTAssertEqual(file.type, "jpg")

        XCTAssertEqual(file.allResponseFields as NSDictionary, response as NSDictionary)
    }
}
